flocker/ca/functional/test_validation.py

```diff
@@ -318,3 +330,62 @@ class RESTAPIContextFactoryValidationTests(
     """
     Tests for the context factory that validates REST API clients.
     """
+
+
+class RequestsTests(TestCase):
+    """
+    Tests for the ``requests`` library's ability to talk to REST API TLS
+    server.
+    """
+    def assert_requests_validates(self, hostname, cert_name):
+        """
+        Assert that ``requests`` succeeds in doing a TLS query to our server.
+
+        :param hostname: Hostname to connect to.
+        :param cert_name: Which attribute of the ``ControlSet`` to use.
+
+        :return: ``Deferred`` firing when test is done.
+        """
+        _, port = find_free_port()
+        ca_set, _ = get_credential_sets()
+        context_factory = rest_api_context_factory(
+            ca_certificate=ca_set.root.credential.certificate,
+            control_credential=getattr(ca_set, cert_name))
+        start_tls_server(self, port, context_factory)
+
+        certs = FilePath(self.mktemp())
+        certs.makedirs()
+        ca_set.copy_to(certs, user=True)
+
+        def req():
+            return requests.get(
+                "https://{}:{}/".format(hostname, port),
+                headers={"connection": "close"},
+                stream=False,
+                cert=(certs.child(b"user.crt").path,
+                      certs.child(b"user.key").path),
+                verify=certs.child(b"cluster.crt").path).content
+        pool = ThreadPool()
+        pool.start()
+        self.addCleanup(pool.stop)
+        d = deferToThreadPool(reactor, pool, req)
+        d.addCallback(self.assertEqual, EXPECTED_STRING)
+        return d
```

Comment:
It would be nice to have the most possible information when this test fails.  If I comment out the `verify` argument in the `requests.get` call above then I do see a nice (well, probably as nice as feasible) traceback ending with an `SSLError`.  If I comment out the `cert` section, though, then the TLS error is on the server side and doesn't seem to get logged at all.  An `SSLError` still shows up on the client but it's not nearly as informative as the first one.

Should we err on the side of logging too much about TLS errors in our server setup?  And should we do the same thing for the actual HTTPS server (slightly sad this test can't/doesn't share the relevant code with the actual HTTPS server...)?  Can we even do that without hacking through klein and Twisted Web? :/

I would expect that people trying out their own clusters and getting stuck on TLS problems might appreciate as much information as possible (or, as usual, we might appreciate it when they come asking us for help).  Making test failures easier to debug seems like a good thing.  Making real-world failures easier to debug seems like a _really_ good thing.